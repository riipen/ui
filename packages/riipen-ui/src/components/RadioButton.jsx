import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";
import _JSXStyle from "styled-jsx/style";

import ThemeContext from "../styles/ThemeContext";
import withClasses from "../utils/withClasses";

import Typography from "./Typography";

const RadioButton = ({
  checked,
  classes,
  color,
  disabled,
  label,
  id,
  prefix,
  size,
  suffix,
  ...other
}) => {
  const theme = React.useContext(ThemeContext);

  let typographyVariant;
  if (size === "small") {
    typographyVariant = "body2";
  } else if (size === "large") {
    typographyVariant = "h5";
  }

  return (
    <div
      className={clsx(
        classes,
        color,
        checked && "checked",
        disabled && "disabled",
        "root",
        size
      )}
    >
      <label
        className={clsx(checked && "checked", disabled && "disabled")}
        htmlFor={id}
      >
        <input checked={checked} disabled={disabled} type="radio" {...other} />
        <Typography
          component="span"
          variant={typographyVariant}
          color="inherit"
        >
          {prefix && <span className="content">{prefix}</span>}
          <span className="content">{label}</span>
          {suffix && <span className="content">{suffix}</span>}
        </Typography>
      </label>
      <style jsx>{`
        /* Common */

        .content {
          margin: 0 5px;
        }

        .root {
          display: inline-flex;
          cursor: pointer;
          border: 1px solid ${theme.palette.grey[300]};
          transition: 0.2s linear;
        }

        label {
          cursor: pointer;
          text-align: center;
        }

        input {
          opacity: 0;
          position: absolute;
          pointer-events: none;
        }

        .disabled {
          cursor: default;
        }

        /* Sizes */

        .small label {
          min-width: 100px;
          padding: 4px;
          width: 100px;
        }

        .medium label {
          min-width: 150px;
          padding: 8px;
          width: 150px;
        }

        .large label {
          min-width: 200px;
          padding: 10px;
          width: 200px;
        }

        /* Colors */

        .root {
          background-color: transparent;
          border-color: ${theme.palette.grey[300]};
          color: ${theme.palette.text.secondary};
        }

        .root.default:hover {
          border-color: ${theme.palette.grey[600]};
          color: ${theme.palette.text.primary};
        }

        .root.default.checked {
          background-color: ${theme.palette.grey[300]};
          border-color: ${theme.palette.grey[300]};
          color: ${theme.palette.common.black};
        }

        .root.primary:hover {
          border-color: ${theme.palette.primary.light};
          color: ${theme.palette.primary.dark};
        }

        .root.primary.checked {
          background-color: ${theme.palette.primary.main};
          border-color: ${theme.palette.primary.main};
          color: ${theme.palette.common.white};
        }

        .root.secondary:hover {
          border-color: ${theme.palette.secondary.light};
          color: ${theme.palette.secondary.dark};
        }

        .root.secondary.checked {
          background-color: ${theme.palette.secondary.main};
          border-color: ${theme.palette.secondary.main};
          color: ${theme.palette.common.white};
        }

        .root.tertiary:hover {
          border-color: ${theme.palette.tertiary.light};
          color: ${theme.palette.tertiary.dark};
        }

        .root.tertiary.checked {
          background-color: ${theme.palette.tertiary.main};
          border-color: ${theme.palette.tertiary.main};
          color: ${theme.palette.common.white};
        }

        .root.disabled:hover,
        .root.disabled {
          background-color: ${theme.palette.disabled};
          border-color: ${theme.palette.disabled};
          color: ${theme.palette.grey[200]};
        }
      `}</style>
    </div>
  );
};

RadioButton.propTypes = {
  /**
   * If `true`, the component is checked.
   */
  checked: PropTypes.bool,

  /**
   * An array of custom CSS classes to apply.
   */
  classes: PropTypes.array,

  /**
   * The color of the radio button.
   */
  color: PropTypes.oneOf(["default", "primary", "secondary", "tertiary"]),

  /**
   * If `true`, the radio will be disabled.
   */
  disabled: PropTypes.bool,

  /**
   * The id of the input component.
   */
  id: PropTypes.string,

  /**
   * Label text to display for the radio.
   */
  label: PropTypes.node,

  /**
   * The component to render before the label text.
   */
  prefix: PropTypes.node,

  /**
   * The size of the radio button
   */
  size: PropTypes.oneOf(["small", "medium", "large"]),

  /**
   * The component to render after the label text.
   */
  suffix: PropTypes.node
};

RadioButton.defaultProps = {
  size: "medium",
  color: "default"
};

RadioButton.displayName = "RadioButton";

export default withClasses(RadioButton);
