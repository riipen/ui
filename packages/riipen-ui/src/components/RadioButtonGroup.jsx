import css from "styled-jsx/css";
import PropTypes from "prop-types";
import React from "react";
import _JSXStyle from "styled-jsx/style";

import clsx from "clsx";

import withClasses from "../utils/withClasses";
import withThemeContext from "../utils/withThemeContext";

import InputLabel from "./InputLabel";
import Typography from "./Typography";

const RadioButtonGroup = ({
  children,
  error,
  hint,
  id,
  label,
  labelProps,
  onChange,
  required,
  theme,
  value,
  variant,
  warning,
  ...other
}) => {
  const linkedStyles = css.resolve`
    .connected.radioButtons .radioButton:first-child {
      border-top-left-radius: ${theme.shape.borderRadius.md};
      border-bottom-left-radius: ${theme.shape.borderRadius.md};
    }

    .connected.radioButtons .radioButton:last-child {
      border-top-right-radius: ${theme.shape.borderRadius.md};
      border-bottom-right-radius: ${theme.shape.borderRadius.md};
    }

    .connected..radioButtons .radioButton:not(:first-child) {
      border-left-color: transparent;
    }

    .connected.radioButtons .radioButton:not(:first-child .checked):hover {
      border-left-color: inherit;
    }

    .separate.radioButtons .radioButton {
      border-radius: ${theme.shape.borderRadius.md};
    }

    .separate.radioButtons .radioButton:not(:last-child) {
      margin-right: ${theme.spacing(4)}px;
    }
  `;

  const handleChange = e => {
    if (onChange) {
      onChange(e, e.target.value);
    }
  };

  const childrenWithProps = React.Children.map(
    children,
    child =>
      child &&
      React.cloneElement(child, {
        checked: value === child.props.value,
        onChange: handleChange,
        ...other,
        ...child.props,
        classes: [
          linkedStyles.className,
          "radioButton",
          ...(child.props.classes || [])
        ]
      })
  ).filter(Boolean);

  return (
    <>
      <fieldset id={id}>
        {(label || hint) && (
          <InputLabel
            hint={hint}
            htmlFor={other.name}
            required={required}
            {...labelProps}
          >
            {label}
          </InputLabel>
        )}

        <div className={clsx("radioButtons", variant, linkedStyles.className)}>
          {childrenWithProps}
        </div>

        {error && (
          <Typography color="negative" variant="body2">
            {error}
          </Typography>
        )}
        {warning && (
          <Typography color="secondary" variant="body2">
            {warning}
          </Typography>
        )}
      </fieldset>
      {linkedStyles.styles}
      <style jsx>{`
        fieldset {
          border: none;
          margin: 0;
          padding: 0;
        }
      `}</style>
    </>
  );
};

RadioButtonGroup.displayName = "RadioButtonGroup";

RadioButtonGroup.propTypes = {
  /**
   * The content of the component.
   */
  children: PropTypes.node.isRequired,

  /**
   * An error to display below the input.
   */
  error: PropTypes.node,

  /**
   * Hint text to display under the label of the input.
   */
  hint: PropTypes.node,

  /**
   * An ID to set on this component.
   */
  id: PropTypes.string,

  /**
   * Label text to display for the input.
   */
  label: PropTypes.node,

  /**
   * Any props to pass through to the InputLabel
   */
  labelProps: PropTypes.object,

  /**
   * Callback fired when a radio button is selected.
   *
   * @param {object} event The event source of the callback.
   * You can pull out the new value by accessing `event.target.value` (string).
   */
  onChange: PropTypes.func,

  /**
   * If true, an asterisk will be appended to the end of the label.
   */
  required: PropTypes.bool,

  /**
   * @ignore
   * The theme context object
   */
  theme: PropTypes.object,

  /**
   * The value of the selected radio button
   */
  value: PropTypes.any,

  /**
   * Radio button group variant to display
   */
  variant: PropTypes.oneOf(["connected", "separate"]),

  /**
   * A warning to display below the input.
   */
  warning: PropTypes.node
};

RadioButtonGroup.defaultProps = {
  variant: "connected"
};

export default withThemeContext(withClasses(RadioButtonGroup));
