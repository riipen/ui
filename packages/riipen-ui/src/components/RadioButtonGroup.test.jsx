import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";

import RadioButtonGroup from "./RadioButtonGroup";
import RadioButton from "./RadioButton";

describe("<RadioButtonGroup>", () => {
  it("renders without errors", () => {
    let error;
    const child = <RadioButton />;

    try {
      mount(<RadioButtonGroup>{child}</RadioButtonGroup>);
    } catch (e) {
      error = e;
    }

    expect(error).toEqual(undefined);
  });

  it("renders correct snapshot", () => {
    const child = <RadioButton />;

    const wrapper = mount(<RadioButtonGroup>{child}</RadioButtonGroup>);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe("children prop", () => {
    it("renders non null children", () => {
      const child = <RadioButton id="one" />;

      const wrapper = mount(
        <RadioButtonGroup>
          {null}
          {child}
        </RadioButtonGroup>
      );

      const childrenNodes = wrapper.find(".radioButtons").children();

      expect(childrenNodes.at(0).props().id).toEqual("one");
    });
  });

  describe("error prop", () => {
    it("displays error", () => {
      const error = <span>an error!</span>;
      const child = <RadioButton id="one" />;

      const wrapper = mount(
        <RadioButtonGroup error={error}>{child}</RadioButtonGroup>
      );

      expect(wrapper.find("fieldset").contains(error)).toEqual(true);
    });
  });

  describe("hint prop", () => {
    it("renders InputLabel when hint is provided", () => {
      const hint = "it's a hint";
      const child = <RadioButton id="one" />;

      const wrapper = mount(
        <RadioButtonGroup hint={hint}>{child}</RadioButtonGroup>
      );

      expect(wrapper.find("InputLabel")).toHaveLength(1);
    });

    it("passes given hint to InputLabel", () => {
      const hint = "it's a hint";
      const child = <RadioButton id="one" />;

      const wrapper = mount(
        <RadioButtonGroup hint={hint}>{child}</RadioButtonGroup>
      );

      expect(wrapper.find("InputLabel").props().hint).toEqual(hint);
    });
  });

  describe("label prop", () => {
    it("passes label to InputLabel as children", () => {
      const label = "it's a label";
      const child = <RadioButton id="one" />;

      const wrapper = mount(
        <RadioButtonGroup label={label}>{child}</RadioButtonGroup>
      );

      expect(wrapper.find("InputLabel").props().children).toEqual(label);
    });

    it("does not render InputLabel when both label and hint are not provided", () => {
      const child = <RadioButton id="one" />;

      const wrapper = mount(<RadioButtonGroup>{child}</RadioButtonGroup>);

      expect(wrapper.find("InputLabel")).toHaveLength(0);
    });
  });

  describe("name prop", () => {
    it("passes given name to InputLabel as htmlFor prop", () => {
      const name = "hello";
      const label = "it's a label";
      const child = <RadioButton id="one" />;

      const wrapper = mount(
        <RadioButtonGroup name={name} label={label}>
          {child}
        </RadioButtonGroup>
      );

      expect(wrapper.find("InputLabel").props().htmlFor).toEqual(name);
    });
  });

  describe("onChange prop", () => {
    it("sets onChange prop of children nodes", () => {
      const child = <RadioButton id="one" />;

      const wrapper = mount(<RadioButtonGroup>{child}</RadioButtonGroup>);

      expect(wrapper.find("RadioButton").props().onChange).toBeDefined();
    });

    it("invokes handler when child RadioButton is changed", () => {
      const child = <RadioButton id="one" />;
      const onChange = jest.fn();

      const wrapper = mount(
        <RadioButtonGroup onChange={onChange}>{child}</RadioButtonGroup>
      );

      wrapper
        .find("RadioButton")
        .find("input")
        .invoke("onChange")({ target: { value: "a" } });

      expect(onChange).toHaveBeenCalledTimes(1);
    });
  });

  describe("required prop", () => {
    it("passes required prop to InputLabel", () => {
      const child = <RadioButton id="one" />;
      const required = true;
      const label = "it's a label";

      const wrapper = mount(
        <RadioButtonGroup required={required} label={label}>
          {child}
        </RadioButtonGroup>
      );

      expect(wrapper.find("InputLabel").props().required).toEqual(required);
    });
  });

  describe("value prop", () => {
    it("sets the childs checked prop according to the value", () => {
      const child = <RadioButton value="one" id="one" />;

      const wrapper = mount(
        <RadioButtonGroup value="one">{child}</RadioButtonGroup>
      );

      expect(wrapper.find("RadioButton").props().checked).toEqual(true);
    });
  });

  describe("variant prop", () => {
    it("sets the connected variant class when variant is connected", () => {
      const child = <RadioButton value="one" id="one" />;

      const wrapper = mount(
        <RadioButtonGroup variant="connected">{child}</RadioButtonGroup>
      );

      expect(wrapper.find(".radioButtons.connected").length).toEqual(1);
    });

    it("sets the separate variant class when variant is separate", () => {
      const child = <RadioButton value="one" id="one" />;

      const wrapper = mount(
        <RadioButtonGroup variant="separate">{child}</RadioButtonGroup>
      );

      expect(wrapper.find(".radioButtons.separate").length).toEqual(1);
    });
  });

  describe("warning prop", () => {
    it("displays warning", () => {
      const warning = "oh no";
      const child = <RadioButton id="one" />;

      const wrapper = mount(
        <RadioButtonGroup warning={warning}>{child}</RadioButtonGroup>
      );

      expect(wrapper.find("fieldset").contains(warning)).toEqual(true);
    });
  });
});
