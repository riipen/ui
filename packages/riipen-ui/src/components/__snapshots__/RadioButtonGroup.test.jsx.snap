// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RadioButtonGroup> renders correct snapshot 1`] = `
<withClasses(RadioButtonGroup)>
  <RadioButtonGroup
    classes={
      Array [
        "riipen",
        "riipen-radiobuttongroup",
      ]
    }
  >
    <fieldset
      className="jsx-3241897731"
    >
      <span
        className="jsx-3241897731 radioButtons jsx-3248853724"
      >
        <withClasses(RadioButton)
          classes={
            Array [
              "jsx-3248853724",
              "radioButton",
            ]
          }
          key=".0"
          onChange={[Function]}
        >
          <RadioButton
            classes={
              Array [
                "riipen",
                "riipen-radiobutton",
                "jsx-3248853724",
                "radioButton",
              ]
            }
            onChange={[Function]}
          >
            <div
              className="jsx-1306841623 riipen riipen-radiobutton jsx-3248853724 radioButton root"
            >
              <label
                className="jsx-1306841623 "
              >
                <input
                  className="jsx-1306841623 "
                  onChange={[Function]}
                  type="radio"
                />
                <withClasses(Typography)
                  color="grey600"
                  variant="body2"
                >
                  <Typography
                    classes={
                      Array [
                        "riipen",
                        "riipen-typography",
                      ]
                    }
                    color="grey600"
                    display="initial"
                    gutter={false}
                    textAlign="inherit"
                    textTransform="inherit"
                    variant="body2"
                  >
                    <p
                      className="jsx-3017984631 root color-grey600 initial body2 align-inherit transform-inherit riipen riipen-typography"
                    >
                      <span
                        className="jsx-1306841623 content"
                      />
                    </p>
                    <JSXStyle
                      dynamic={
                        Array [
                          "#424242",
                          "Roboto, Helvetica, Arial, sans-serif",
                          400,
                          "14px",
                          "20px",
                          "0.00938em",
                          "#424242",
                          "Roboto, Helvetica, Arial, sans-serif",
                          400,
                          "12px",
                          "17px",
                          "0.01071em",
                          "#424242",
                          "Roboto, Helvetica, Arial, sans-serif",
                          400,
                          "10px",
                          "14px",
                          "0.01071em",
                          "#373737",
                          "Roboto, Helvetica, Arial, sans-serif",
                          500,
                          "40px",
                          "56px",
                          "-0.01562em",
                          "#373737",
                          "Roboto, Helvetica, Arial, sans-serif",
                          500,
                          "30px",
                          "42px",
                          "-0.00833em",
                          "#373737",
                          "Roboto, Helvetica, Arial, sans-serif",
                          500,
                          "22px",
                          "32px",
                          "0em",
                          "#373737",
                          "Roboto, Helvetica, Arial, sans-serif",
                          400,
                          "20px",
                          "30px",
                          "-0.00735em",
                          "#373737",
                          "Roboto, Helvetica, Arial, sans-serif",
                          400,
                          "16px",
                          "23px",
                          "0rem",
                          "#000",
                          "#747474",
                          "#424242",
                          "#373737",
                          "#1f71ae",
                          "#f58632",
                          "#3caabb",
                          "#4cc88d",
                          "#e95353",
                          "#fff",
                          700,
                          300,
                          500,
                          400,
                        ]
                      }
                      id="3990007342"
                    />
                  </Typography>
                </withClasses(Typography)>
              </label>
              <JSXStyle
                dynamic={
                  Array [
                    "#dddddd",
                    "#eeeeee",
                  ]
                }
                id="547299733"
              />
            </div>
          </RadioButton>
        </withClasses(RadioButton)>
      </span>
    </fieldset>
    <JSXStyle
      id="3248853724"
    />
    <JSXStyle
      id="3241897731"
    />
  </RadioButtonGroup>
</withClasses(RadioButtonGroup)>
`;
